### ISSUE TITLE
Issue: <Please tell me how to combine Routing and RAG in a chain.>

### ISSUE THREAD

**Issue by kaeruko** [langchain-ai/langchain#15193]
### Issue you'd like to raise.

I apologize for the naive question.it's not about an error or a bug.
I'm trying to implement routing by following the guide here: https://python.langchain.com/docs/modules/chains/foundational/router

However, I can't figure out how to use RAG.
I tried changing the last code in the guide like this.

```python
final_chain = (
    RunnablePassthrough.assign(topic=itemgetter("input") | classifier_chain)
    | prompt_branch
    | ChatOpenAI()
    | StrOutputParser()
)
```

```python
final_chain = (
    {
        "context": retriever,
        "topic": itemgetter("input") | classifier_chain,
    }
    | prompt_branch
    | llm
    | StrOutputParser()
)
```
But I get the following error:

```shell
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/tiktoken/core.py", line 116, in encode
    if match := _special_token_regex(disallowed_special).search(text):
TypeError: expected string or buffer
```

**Comment by kaeruko** [comment]
Thank you! I tried it, but I'm still getting the same error...
I'll study while checking the source code you provided.

```python
final_chain = (
    RunnableParallel({"context": retriever, "question": RunnablePassthrough()})
    | prompt_branch
    | llm
    | StrOutputParser()
)
final_chain.invoke(
    {
        "input": query
    }
)
```

got a error
```bash
Traceback (most recent call last):
  File "/Users/user/work/python/new-service-datasets/./script/langchain_router.py", line 263, in <module>
    pipeline.invoke(
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_core/runnables/base.py", line 1514, in invoke
    input = step.invoke(
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_core/runnables/base.py", line 2040, in invoke
    output = {key: future.result() for key, future in zip(steps, futures)}
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_core/runnables/base.py", line 2040, in <dictcomp>
    output = {key: future.result() for key, future in zip(steps, futures)}
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/concurrent/futures/_base.py", line 438, in result
    return self.__get_result()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_core/retrievers.py", line 112, in invoke
    return self.get_relevant_documents(
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_core/retrievers.py", line 211, in get_relevant_documents
    raise e
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_core/retrievers.py", line 204, in get_relevant_documents
    result = self._get_relevant_documents(
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain/retrievers/multi_vector.py", line 65, in _get_relevant_documents
    sub_docs = self.vectorstore.similarity_search(query, **self.search_kwargs)
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_community/vectorstores/faiss.py", line 512, in similarity_search
    docs_and_scores = self.similarity_search_with_score(
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_community/vectorstores/faiss.py", line 393, in similarity_search_with_score
    embedding = self._embed_query(query)
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_community/vectorstores/faiss.py", line 156, in _embed_query
    return self.embedding_function.embed_query(text)
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_community/embeddings/openai.py", line 696, in embed_query
    return self.embed_documents([text])[0]
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_community/embeddings/openai.py", line 667, in embed_documents
    return self._get_len_safe_embeddings(texts, engine=engine)
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/langchain_community/embeddings/openai.py", line 470, in _get_len_safe_embeddings
    token = encoding.encode(
  File "/Users/user/Library/Python/3.9/lib/python/site-packages/tiktoken/core.py", line 116, in encode
    if match := _special_token_regex(disallowed_special).search(text):
TypeError: expected string or buffer
```

**Comment by austinbv** [comment]
The issue comes from the full dictionary getting passed into the embedder. You can fix it with

```py
final_chain = (
    {
        "context": itemgetter("input") | retriever, # <---------------- Pass just the input into the retriever 
        "topic": itemgetter("input") | classifier_chain,
    }
    | prompt_branch
    | llm
    | StrOutputParser()
)
```

**Comment by kaeruko** [comment]
Thank you for your kindness. 
I'm encountering a different error now, but it seems to be caused by something outside of the code I shared in this issue, so I'm working hard on debugging it. 
If I find myself completely stuck, I might reach out for your help again.

**Comment by kaeruko** [comment]
If I should have created a new issue instead of reopening a closed one, I apologize.

Following your kind advice, I tried the suggested method and thankfully, the error has been resolved. However, it seems that I'm not getting any response from 
`pipeline.invoke({"input":query})`.

Here is my code: https://github.com/kaeruko/plactis/blob/master/langchain_test.py#L206-L215

When I dump the contents of `call_func_with_variable_args`, the `context` appears to be empty, and it seems retrieval is not happening. I would like to check the retrieval process of langchain, but I'm having trouble identifying where that's happening in the stack trace..."
https://github.com/kaeruko/plactis/blob/master/langchain_test.py#L206-L215
